src/PrMerger.js

```diff
@@ -43,12 +44,31 @@ class PrMerger {
     // there is a PR still-in-merge.
     async runStep() {
         Logger.info("runStep running");
-        const currentContext = await this._current();
-        if (currentContext && !(await this._finishContext(currentContext)))
-            return true; // still in-process
 
-        const prList = await this._getPRList();
-        this._logPRList(prList);
+        let prList = await this._getPRList();
+
+        let currentContext = await this._current();
+
+        if (currentContext && Config.guardedRun()) {
+            const clearedForMerge = await this._clearedForMerge(currentContext.number());
+            if (!clearedForMerge && prList.length && prList[0].clearedForMerge) {
+                Logger.info("will try to give up the not-cleared-for-merge PR" + currentContext.number() +
+                        " in favor of a cleared-for-merge PR" + prList[0].number);
+                // Give only other PRs a chance to start.
+                // Without removing current PR from the list, we can get into
+                // a loop of useless restarting it from scratch.
```

Comment:
If we can get into a loop, then there is a bug somewhere that we should fix instead AFAICT. The situation is simple: When there are labeled PRs, we abort any current merging of an unlabeled PR and do not start merging of any unlabeled PR.

If it makes things easier, we do not have to start merging any unlabeled PR _regardless_ of whether labeled PRs exist or not. In other words, we only start merging labeled PRs.

All of the above applies to guarded runs only, of course.

Reply:
> If we can get into a loop

We could get into the loop without my changes. With my solution, the bot switches from the unlabeled PR to a labeled only if such labeled PR has passed all checks because there can be many labeled, but defective PRs. If all labeled PRs are invalid yet, the bot keeps hanging on the unlabeled PR (and waiting until it becomes labeled).
With your (1) solution we would delete staged commit for our unlabeled PR and would not start any other PR. With your (2) solution we would not even start the unlabeled PR at all.
I think my solution is better because it works for more cases (bot processes more PRs) while it does not have disadvantages, IMO.